packages/cojson/src/PeerState.test.ts

```diff
@@ -0,0 +1,92 @@
+        expect(peerState.crashOnClose).toBe(false);
+        expect(peerState.closed).toBe(false);
+        expect(peerState.optimisticKnownStates).toEqual({});
+        expect(peerState.toldKnownState).toEqual(new Set());
+    });
+
+    test("should push outgoing message to peer", async () => {
+        const { mockPeer, peerState } = setup();
+        const message: SyncMessage = { action: "load", id: "co_ztest-id", header: false, sessions: {} };
+        await peerState.pushOutgoingMessage(message);
+        expect(mockPeer.outgoing.push).toHaveBeenCalledWith(message);
+    });
+
+    test("should return peer's incoming when not closed", () => {
+        const { mockPeer, peerState } = setup();
+        expect(peerState.incoming).toBe(mockPeer.incoming);
+    });
+
+    test("should return Disconnected when closed", async () => {
+        const { peerState } = setup();
+        peerState.gracefulShutdown();
+        const incomingIterator = peerState.incoming[Symbol.asyncIterator]();
+        const { value, done } = await incomingIterator.next();
+        expect(value).toBe("Disconnected");
+        expect(done).toBe(false);
+    });
+
+    test("should perform graceful shutdown", () => {
+        const { mockPeer, peerState } = setup();
+        const consoleSpy = vi.spyOn(console, "debug").mockImplementation(() => {});
+        peerState.gracefulShutdown();
+        expect(mockPeer.outgoing.close).toHaveBeenCalled();
+        expect(peerState.closed).toBe(true);
+        expect(consoleSpy).toHaveBeenCalledWith("Gracefully closing", "test-peer");
+        consoleSpy.mockRestore();
+    });
+
+    test("should schedule outgoing messages based on their priority", async () => {
+        const { peerState } = setup();
+
+        const loadMessage: SyncMessage = { action: "load", id: "co_zhigh", header: false, sessions: {} };
+        const contentMessageHigh: SyncMessage = { action: "content", id: "co_zhigh", new: {}, priority: CO_VALUE_PRIORITY.HIGH };
+        const contentMessageMid: SyncMessage = { action: "content", id: "co_zmid", new: {}, priority: CO_VALUE_PRIORITY.MEDIUM };
+        const contentMessageLow: SyncMessage = { action: "content", id: "co_zlow", new: {}, priority: CO_VALUE_PRIORITY.LOW };
+
+        const promises = [
+            peerState.pushOutgoingMessage(contentMessageLow),
+            peerState.pushOutgoingMessage(contentMessageMid),
+            peerState.pushOutgoingMessage(contentMessageHigh),
+            peerState.pushOutgoingMessage(loadMessage),
+        ];
+
+        await Promise.all(promises);
+
+        // The first message is pushed directly, the other three are queued because are waiting
+        // for the first push to be completed.
+        expect(peerState["peer"].outgoing.push).toHaveBeenNthCalledWith(1, contentMessageLow); 
+
+        // Load message are managed as high priority messages and having the same priority as the content message
+        // they follow the push order.
```

Comment:
@aeplay 

I'm considering all the non-content messages as HIGH priority.
Not sure which effect this might have on the system.

My guess was that by giving to the "exchange" message a higher priority we ensure that is processed before any kind of content upload (with the exception of Groups and Accounts) reducing the delay on critical actions like "invitations".

Reply:
yes, this is good and correct. State info is always more important than content